File: internal/backend/llvm/math.go
package llvm

import (
	"fmt"
	"strings"

	"github.com/pilotworks/scriptgo/internal/ir"
)

func emitMathIntrinsic(out *strings.Builder, instruction ir.Instruction) error {
	if instruction.Type != ir.TypeNumber {
		return fmt.Errorf("math intrinsic %q has invalid signature", instruction.Callee)
	}
	switch instruction.Callee {
	case "__Math.abs", "__Math.ceil", "__Math.floor", "__Math.trunc", "__Math.sqrt", "__Math.round", "__Math.sin", "__Math.cos", "__Math.log", "__Math.log2", "__Math.log10", "__Math.exp", "__Math.tan", "__Math.atan", "__Math.asin", "__Math.acos", "__Math.cbrt", "__Math.sinh", "__Math.cosh", "__Math.tanh", "__Math.asinh", "__Math.acosh", "__Math.atanh", "__Math.expm1", "__Math.log1p":
		if len(instruction.Args) != 1 {
			return fmt.Errorf("math intrinsic %q requires one number", instruction.Callee)
		}
		name := map[string]string{
			"__Math.abs":   "llvm.fabs.f64",
			"__Math.ceil":  "llvm.ceil.f64",
			"__Math.floor": "llvm.floor.f64",
			"__Math.trunc": "llvm.trunc.f64",
			"__Math.sqrt":  "llvm.sqrt.f64",
			"__Math.round": "llvm.round.f64",
			"__Math.sin":   "llvm.sin.f64",
			"__Math.cos":   "llvm.cos.f64",
			"__Math.log":   "llvm.log.f64",
			"__Math.log2":  "llvm.log2.f64",
			"__Math.log10": "llvm.log10.f64",
			"__Math.exp":   "llvm.exp.f64",
			"__Math.tan":   "tan",
			"__Math.atan":  "atan",
			"__Math.asin":  "asin",
			"__Math.acos":  "acos",
			"__Math.cbrt":  "cbrt",
			"__Math.sinh":  "sinh",
			"__Math.cosh":  "cosh",
			"__Math.tanh":  "tanh",
			"__Math.asinh": "asinh",
			"__Math.acosh": "acosh",
			"__Math.atanh": "atanh",
			"__Math.expm1": "expm1",
			"__Math.log1p": "log1p",
		}[instruction.Callee]
		fmt.Fprintf(out, "  %%%s = call double @%s(double %%%s)\n", instruction.Result, name, instruction.Args[0])
		return nil
	case "__Math.fround":
		if len(instruction.Args) != 1 {
			return fmt.Errorf("math intrinsic %q requires one number", instruction.Callee)
		}
		f32 := instruction.Result + "_f32"
		fmt.Fprintf(out, "  %%%s = fptrunc double %%%s to float\n", f32, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = fpext float %%%s to double\n", instruction.Result, f32)
		return nil
	case "__Math.f16round":
		if len(instruction.Args) != 1 {
			return fmt.Errorf("math intrinsic %q requires one number", instruction.Callee)
		}
		f16 := instruction.Result + "_f16"
		fmt.Fprintf(out, "  %%%s = fptrunc double %%%s to half\n", f16, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = fpext half %%%s to double\n", instruction.Result, f16)
		return nil
	case "__Math.clz32":
		if len(instruction.Args) != 1 {
			return fmt.Errorf("math intrinsic %q requires one number", instruction.Callee)
		}
		u32 := instruction.Result + "_u32"
		clz := instruction.Result + "_clz"
		fmt.Fprintf(out, "  %%%s = fptoui double %%%s to i32\n", u32, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = call i32 @llvm.ctlz.i32(i32 %%%s, i1 false)\n", clz, u32)
		fmt.Fprintf(out, "  %%%s = uitofp i32 %%%s to double\n", instruction.Result, clz)
		return nil
	case "__Math.imul":
		if len(instruction.Args) != 2 {
			return fmt.Errorf("math intrinsic %q requires two numbers", instruction.Callee)
		}
		a32 := instruction.Result + "_a32"
		b32 := instruction.Result + "_b32"
		mul := instruction.Result + "_mul"
		fmt.Fprintf(out, "  %%%s = fptosi double %%%s to i32\n", a32, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = fptosi double %%%s to i32\n", b32, instruction.Args[1])
		fmt.Fprintf(out, "  %%%s = mul i32 %%%s, %%%s\n", mul, a32, b32)
		fmt.Fprintf(out, "  %%%s = sitofp i32 %%%s to double\n", instruction.Result, mul)
		return nil
	case "__Math.sign":
		if len(instruction.Args) != 1 {
			return fmt.Errorf("math intrinsic %q requires one number", instruction.Callee)
		}
		gt := instruction.Result + "_gt"
		lt := instruction.Result + "_lt"
		neg := instruction.Result + "_neg"
		fmt.Fprintf(out, "  %%%s = fcmp ogt double %%%s, 0.0\n", gt, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = fcmp olt double %%%s, 0.0\n", lt, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = select i1 %%%s, double -1.0, double %%%s\n", neg, lt, instruction.Args[0])
		fmt.Fprintf(out, "  %%%s = select i1 %%%s, double 1.0, double %%%s\n", instruction.Result, gt, neg)
		return nil
	case "__Math.random":
		fmt.Fprintf(out, "  %%%s = call double @drand48()\n", instruction.Result)
		return nil
	case "__Math.min", "__Math.max", "__Math.pow", "__Math.atan2", "__Math.hypot":
		if len(instruction.Args) != 2 {
			return fmt.Errorf("math intrinsic %q requires two numbers", instruction.Callee)
		}
		name := map[string]string{
			"__Math.min":   "llvm.minnum.f64",
			"__Math.max":   "llvm.maxnum.f64",
			"__Math.pow":   "llvm.pow.f64",
			"__Math.atan2": "atan2",
			"__Math.hypot": "hypot",
		}[instruction.Callee]
		fmt.Fprintf(out, "  %%%s = call double @%s(double %%%s, double %%%s)\n", instruction.Result, name, instruction.Args[0], instruction.Args[1])
		return nil
	default:
		return fmt.Errorf("unknown math intrinsic %q", instruction.Callee)
	}
}

File: internal/compiler/testdata/corpus/api/math.ts
// ScriptGo Corpus: Math Standard Builtin APIs
// Consolidated test suite with inline assertions.

// @api: math.abs
// @expect: 42
// @expect: 42
console.log(Math.abs(-42)); console.log(Math.abs(42));

// @api: math.atan
// @expect: 0
console.log(Math.atan(0));

// @api: math.atan2
// @expect: 0
console.log(Math.atan2(0, 1));

// @api: math.ceil
// @expect: 5
// @expect: -4
console.log(Math.ceil(4.2)); console.log(Math.ceil(-4.8));

// @api: math.constants
// @api: Math.E
// @api: Math.LN2
// @api: Math.PI
// @expect: true
// @expect: true
// @expect: true
console.log(Math.PI > 3.14); console.log(Math.E > 2.71); console.log(Math.LN2 > 0.69);

// @api: math.cos
// @expect: 1
console.log(Math.cos(0));

// @api: math.exp
// @expect: 1
console.log(Math.exp(0));

// @api: math.floor
// @expect: 4
// @expect: -5
console.log(Math.floor(4.8)); console.log(Math.floor(-4.2));

// @api: math.hypot
// @expect: 5
console.log(Math.hypot(3, 4));

// @api: math.log
// @expect: 1
console.log(Math.log(Math.E));

// @api: math.log10
// @expect: 2
console.log(Math.log10(100));

// @api: math.log2
// @expect: 3
console.log(Math.log2(8));

// @api: math.max
// @expect: 20
// @expect: 5
console.log(Math.max(10, 20)); console.log(Math.max(-5, 5));

// @api: math.min
// @expect: 10
// @expect: -5
console.log(Math.min(10, 20)); console.log(Math.min(-5, 5));

// @api: math.pow
// @expect: 256
// @expect: 27
console.log(Math.pow(2, 8)); console.log(Math.pow(3, 3));

// @api: math.random
// @expect: true
const r_math_random_15 = Math.random(); console.log(r_math_random_15 >= 0 && r_math_random_15 < 1);

// @api: math.round
// @expect: 5
// @expect: 4
console.log(Math.round(4.5)); console.log(Math.round(4.4));

// @api: math.sign
// @expect: -1
// @expect: 1
// @expect: 0
console.log(Math.sign(-10)); console.log(Math.sign(10)); console.log(Math.sign(0));

// @api: math.sin
// @expect: 0
console.log(Math.sin(0));

// @api: math.sqrt
// @expect: 4
// @expect: 1.4142135623730951
console.log(Math.sqrt(16)); console.log(Math.sqrt(2));

// @api: math.tan
// @expect: 0
console.log(Math.tan(0));

// @api: math.trunc
// @expect: 4
// @expect: -4
console.log(Math.trunc(4.9)); console.log(Math.trunc(-4.9));

// @api: math.acos
// @expect: 0
console.log(Math.acos(1));

// @api: math.acosh
// @expect: 0
console.log(Math.acosh(1));

// @api: math.asin
// @expect: 0
console.log(Math.asin(0));

// @api: math.asinh
// @expect: 0
console.log(Math.asinh(0));

// @api: math.atanh
// @expect: 0
console.log(Math.atanh(0));

// @api: math.cbrt
// @expect: 4
// @expect: -2
console.log(Math.cbrt(64)); console.log(Math.cbrt(-8));

// @api: math.clz32
// @expect: 27
// @expect: 32
console.log(Math.clz32(16)); console.log(Math.clz32(0));

// @api: math.cosh
// @expect: 1
console.log(Math.cosh(0));

// @api: math.expm1
// @expect: 0
console.log(Math.expm1(0));

// @api: math.fround
// @expect: 1.5
console.log(Math.fround(1.5));

// @api: math.imul
// @expect: 12
// @expect: -12
console.log(Math.imul(3, 4)); console.log(Math.imul(-3, 4));

// @api: math.log1p
// @expect: 0
console.log(Math.log1p(0));

// @api: math.sinh
// @expect: 0
console.log(Math.sinh(0));

// @api: math.tanh
// @expect: 0
console.log(Math.tanh(0));

// @api: Math.LN10
// @expect: true
console.log(Math.LN10 > 2.3);

// @api: Math.LOG10E
// @expect: true
console.log(Math.LOG10E > 0.43);

// @api: Math.LOG2E
// @expect: true
console.log(Math.LOG2E > 1.44);

// @api: Math.SQRT1_2
// @expect: true
console.log(Math.SQRT1_2 > 0.7);

// @api: Math.SQRT2
// @expect: true
console.log(Math.SQRT2 > 1.41);

// @api: Math.f16round
// @expect: 1.5
console.log(Math.f16round(1.5));

